libmscore/changeMap.cpp

```diff
@@ -0,0 +1,547 @@
+                  // ramps during stage 1.
+                  std::sort(rampEvents.begin(), rampEvents.end(), ChangeMap::compareRampEvents);
+                  for (auto& event : rampEvents) {
+                        insert(tick, event);
+                        }
+                  }
+            }
+      }
+
+//---------------------------------------------------------
+//   cleanupStage1
+///   remove any ramps or fixes that are completely enclosed within other ramps
+//---------------------------------------------------------
+
+void ChangeMap::cleanupStage1()
+      {
+      Fraction currentRampStart = Fraction(-1, 1);    // start point of ramp we're in
+      Fraction currentRampEnd = Fraction(-1, 1);      // end point of ramp we're in
+      Fraction lastFix = Fraction(-1, 1);             // the position of the last fix event
+      bool inRamp = false;                            // whether we're in a ramp or not
+
+      // Keep a record of the endpoints
+      EndPointsVector endPoints;
+      std::vector<bool> startsInRamp;
+
+      auto i = begin();
+      while (i != end()) {
+            Fraction tick = i.key();
+            ChangeEvent& event = i.value();
+            Fraction etick = tick + event.length;
+
+            // Reset if we've left the ramp we were in
+            if (currentRampEnd < tick)
+                  inRamp = false;
+
+            if (event.type == ChangeEventType::RAMP) {
+                  if (inRamp) {
+                        if (etick <= currentRampEnd) {
+                              // delete, this event is enveloped
+                              i = erase(i);
+                              // don't add to the end points
+                              continue;
+                              }
+                        else {
+                              currentRampStart = tick;
+                              currentRampEnd = etick;
+                              startsInRamp.push_back(true);
+                              }
+                        }
+                  else {
+                        currentRampStart = tick;
+                        currentRampEnd = etick;
+                        inRamp = true;
+                        startsInRamp.push_back(false);
+                        }
+
+                  endPoints.push_back(std::make_pair(tick, etick));
+                  }
+            else if (event.type == ChangeEventType::FIX) {
+                  if (inRamp) {
+                        if (tick != currentRampStart && tick != currentRampEnd && lastFix != tick) {
+                              // delete, this event is enveloped or at the same point as another fix
+                              i = erase(i);
+                              continue;
+                              }
+                        }
+
+                  lastFix = tick;
+                  }
+
+            i++;
+            }
+
+      cleanupStage2(startsInRamp, endPoints);
```

Comment:
It would be more logical to call `cleanupStage2()` from the same place where `cleanupStage1()` is called. As an alternative, `cleanupStage2()` could just be given a separate name.